File: MobileAppReactNativeMobX/src/business/state/Stock.ts
import { makeAutoObservable } from 'mobx';

/** Stocks the user owns. */
class Stock {
  readonly ticker: string;
  howManyShares: number;
  averagePrice: number;

  constructor(ticker: string, howManyShares: number, averagePrice: number) {
    this.ticker = ticker;
    this.howManyShares = howManyShares;
    this.averagePrice = averagePrice;
    makeAutoObservable(this);
  }

  get costBasis() {
    return this.howManyShares * this.averagePrice;
  }

  get averagePriceStr() {
    return `US$ ${this.averagePrice.toFixed(2)}`;
  }

  toString(): string {
    return `${this.ticker}: ${this.howManyShares} shares`;
  }
}

export default Stock;

File: MobileAppFlutterRedux/lib/models/stock.dart
import 'package:flutter/foundation.dart';

import 'utils/map_deserialization_extension.dart';

/// Stocks the user has.
@immutable
class Stock {
  final String ticker;
  final int howManyShares;
  final double averagePrice;

  Stock({
    required this.ticker,
    required this.howManyShares,
    required this.averagePrice,
  });

  double get costBasis => howManyShares * averagePrice;

  String get averagePriceStr => 'US\$ ${averagePrice.toStringAsFixed(2)}';

  @override
  String toString() => '$howManyShares $ticker @$averagePrice';

  Map<String, dynamic> toJson() => {
        'ticker': ticker,
        'howManyShares': howManyShares,
        'averagePrice': averagePrice,
      };

  static Stock fromJson(Json json) {
    return Stock(
      ticker: json.asString('ticker')!,
      howManyShares: json.asInt('howManyShares')!,
      averagePrice: json.asDouble('averagePrice')!,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Stock &&
          runtimeType == other.runtimeType &&
          ticker == other.ticker &&
          howManyShares == other.howManyShares &&
          averagePrice == other.averagePrice;

  @override
  int get hashCode => ticker.hashCode ^ howManyShares.hashCode ^ averagePrice.hashCode;
}
